File: WinForms/CS/GridMainDemo/ServerMode/GridInstantFeedbackUI.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.Xpo.DB;
using DevExpress.Xpo;
using DevExpress.Xpo.Helpers;
using System.Diagnostics;
using DevExpress.XtraEditors.Controls;
using DevExpress.Utils;
using DevExpress.DXperience.Demos;

namespace DevExpress.XtraGrid.Demos {
    public partial class GridInstantFeedbackUI : DevExpress.XtraGrid.Demos.TutorialControl {
        LogWriter writer = new LogWriter();
        public GridInstantFeedbackUI() {
             if(ServerModeStrings.SQLConnectionString == string.Empty)
                ShowConnectionWizard();
            try {
                IDataStore store = XpoDefault.GetConnectionProvider(ServerModeStrings.SQLConnectionString, AutoCreateOption.DatabaseAndSchema);
                store = new WaitCursorWrapper(store);
                store = new DataStoreLogger(store, writer);
                XpoDefault.DataLayer = new SimpleDataLayer(store);
            }
            catch {
                XtraMessageBox.Show(SQLConnectionHelper.FailedConnection, SQLConnectionHelper.FailedConnectionCaption, MessageBoxButtons.OK, MessageBoxIcon.Error);
                ServerModeStrings.SQLConnectionString = "";
            }
            InitializeComponent();
            //<gridControl1>
            /*
            ~Note: the following property is set at design time and listed here only for demonstration purposes.
            ~Assign a data source to the grid via the DataSource property:
            
            gridControl1.DataSource = xpAsyncServerModeSource1;
            
            ~Sort rows by the Sent column in descending order 
            */
            xpAsyncServerModeSource1.DefaultSorting = "Sent desc";
            //</gridControl1>
            colOid.AllowSummaryMenu = false;
            colExpression.AllowSummaryMenu = false;
            colHasAttachment.AllowSummaryMenu = false;
            colOid.OptionsColumn.AllowGroup = DefaultBoolean.False;
            colSize.OptionsColumn.AllowGroup = DefaultBoolean.False;
            InitEditors();
        }
        protected override string[] WhatsThisCodeFileNames {
            get { return new string[] { "GridInstantFeedbackUI" }; }
        }
        protected override string WhatsThisXMLFileName {
            get { return "GridInstantFeedbackUI"; }
        }
        protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            OnStyleChanged();
        }
        void InitEditors() {
            foreach(WaitAnimationOptions mode in Enum.GetValues(typeof(WaitAnimationOptions)))
                icbAnimation.Properties.Items.Add(new ImageComboBoxItem(EnumTitlesKeeper<WaitAnimationOptions>.GetTitle(mode), mode, -1));
            icbAnimation.EditValue = gridView1.OptionsView.WaitAnimationOptions;
        }
        private void ShowConnectionWizard() {
            ShowConnectionWizard(string.Empty);
        }
        private void ShowConnectionWizard(string demoString) {
            frmSQLConnector frm = demoString == string.Empty ? new frmSQLConnector() : new frmSQLConnector(demoString);
            frm.ShowDialog();
            ServerModeStrings.SQLConnectionString = frm.GetDataBaseConnectionString();
        }

        private void sbShowSQLConnection_Click(object sender, EventArgs e) {
            ShowConnectionWizard("Return");
        }

         private void xpAsyncServerModeSource1_ResolveSession(object sender, ResolveSessionEventArgs e) {
            try {
                Session session = new Session();
                session.ConnectionString = ServerModeStrings.SQLConnectionString;
                session.Connect();
                e.Session = session;
                e.Tag = session;
            } catch {
                e.Session = new Session(new SimpleDataLayer(new InMemoryDataStore()));
                e.Tag = null;
            }
        }

        private void xpAsyncServerModeSource1_DismissSession(object sender, ResolveSessionEventArgs e) {
            IDisposable session = e.Tag as IDisposable;
            if(session != null)
                session.Dispose();
        }

        //<icbAnimation>
        private void icbAnimation_SelectedIndexChanged(object sender, EventArgs e) {
            gridView1.OptionsView.WaitAnimationOptions = (WaitAnimationOptions)icbAnimation.EditValue;
        }
        //</icbAnimation>

        private void gridView1_CustomColumnDisplayText(object sender, DevExpress.XtraGrid.Views.Base.CustomColumnDisplayTextEventArgs e) {
            if (e.Column == colExpression) {
                int value = 0;
                if (int.TryParse(string.Format("{0}", e.Value), out value))
                    e.DisplayText = string.Format("{0}:{1:00}", value / 60, value % 60);
            }
        }
    }
}

File: WinForms/VB/GridMainDemo/ServerMode/GridInstantFeedbackUI.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.Xpo.DB
Imports DevExpress.Xpo
Imports DevExpress.Xpo.Helpers
Imports System.Diagnostics
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.Utils
Imports DevExpress.DXperience.Demos

Namespace DevExpress.XtraGrid.Demos

    Public Partial Class GridInstantFeedbackUI
        Inherits TutorialControl

        Private writer As LogWriter = New LogWriter()

        Public Sub New()
            If Equals(ServerModeStrings.SQLConnectionString, String.Empty) Then ShowConnectionWizard()
            Try
                Dim store As IDataStore = XpoDefault.GetConnectionProvider(ServerModeStrings.SQLConnectionString, AutoCreateOption.DatabaseAndSchema)
                store = New WaitCursorWrapper(store)
                store = New DataStoreLogger(store, writer)
                XpoDefault.DataLayer = New SimpleDataLayer(store)
            Catch
                XtraMessageBox.Show(SQLConnectionHelper.FailedConnection, SQLConnectionHelper.FailedConnectionCaption, MessageBoxButtons.OK, MessageBoxIcon.Error)
                ServerModeStrings.SQLConnectionString = ""
            End Try

            InitializeComponent()
            '<gridControl1>
            ' 
            ' ~Note: the following property is set at design time and listed here only for demonstration purposes.
            ' ~Assign a data source to the grid via the DataSource property:
            ' 
            ' gridControl1.DataSource = xpAsyncServerModeSource1;
            ' 
            ' ~Sort rows by the Sent column in descending order 
            ' 
            xpAsyncServerModeSource1.DefaultSorting = "Sent desc"
            '</gridControl1>
            colOid.AllowSummaryMenu = False
            colExpression.AllowSummaryMenu = False
            colHasAttachment.AllowSummaryMenu = False
            colOid.OptionsColumn.AllowGroup = DefaultBoolean.False
            colSize.OptionsColumn.AllowGroup = DefaultBoolean.False
            InitEditors()
        End Sub

        Protected Overrides ReadOnly Property WhatsThisCodeFileNames As String()
            Get
                Return New String() {"GridInstantFeedbackUI"}
            End Get
        End Property

        Protected Overrides ReadOnly Property WhatsThisXMLFileName As String
            Get
                Return "GridInstantFeedbackUI"
            End Get
        End Property

        Protected Overrides Sub OnLoad(ByVal e As EventArgs)
            MyBase.OnLoad(e)
            OnStyleChanged()
        End Sub

        Private Sub InitEditors()
            For Each mode As WaitAnimationOptions In [Enum].GetValues(GetType(WaitAnimationOptions))
                icbAnimation.Properties.Items.Add(New ImageComboBoxItem(EnumTitlesKeeper(Of WaitAnimationOptions).GetTitle(mode), mode, -1))
            Next

            icbAnimation.EditValue = gridView1.OptionsView.WaitAnimationOptions
        End Sub

        Private Sub ShowConnectionWizard()
            ShowConnectionWizard(String.Empty)
        End Sub

        Private Sub ShowConnectionWizard(ByVal demoString As String)
            Dim frm As frmSQLConnector = If(Equals(demoString, String.Empty), New frmSQLConnector(), New frmSQLConnector(demoString))
            frm.ShowDialog()
            ServerModeStrings.SQLConnectionString = frm.GetDataBaseConnectionString()
        End Sub

        Private Sub sbShowSQLConnection_Click(ByVal sender As Object, ByVal e As EventArgs)
            ShowConnectionWizard("Return")
        End Sub

        Private Sub xpAsyncServerModeSource1_ResolveSession(ByVal sender As Object, ByVal e As ResolveSessionEventArgs)
            Try
                Dim session As Session = New Session()
                session.ConnectionString = ServerModeStrings.SQLConnectionString
                session.Connect()
                e.Session = session
                e.Tag = session
            Catch
                e.Session = New Session(New SimpleDataLayer(New InMemoryDataStore()))
                e.Tag = Nothing
            End Try
        End Sub

        Private Sub xpAsyncServerModeSource1_DismissSession(ByVal sender As Object, ByVal e As ResolveSessionEventArgs)
            Dim session As IDisposable = TryCast(e.Tag, IDisposable)
            If session IsNot Nothing Then session.Dispose()
        End Sub

        '<icbAnimation>
        Private Sub icbAnimation_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            gridView1.OptionsView.WaitAnimationOptions = CType(icbAnimation.EditValue, WaitAnimationOptions)
        End Sub

        '</icbAnimation>
        Private Sub gridView1_CustomColumnDisplayText(ByVal sender As Object, ByVal e As Views.Base.CustomColumnDisplayTextEventArgs)
            If e.Column Is colExpression Then
                Dim value As Integer = 0
                If Integer.TryParse(String.Format("{0}", e.Value), value) Then e.DisplayText = String.Format("{0}:{1:00}", value \ 60, value Mod 60)
            End If
        End Sub
    End Class
End Namespace
